firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -548,10 +571,20 @@ public void releaseQuery(Query query) {
 
           int targetId = queryData.getTargetId();
           QueryData cachedQueryData = targetIds.get(targetId);
+
+          boolean needsUpdate = false;
```

Comment:
Given how huge this method already is, consider pulling out this needs-update logic into a helper where you can make the internal decisions with simple returns.

Reply:
I went down this route, but since there is already a `shouldPersistQueryData()` in LocalStore, I am afraid that adding an additional method that is named similarly but doesn't quite do what the existing method does will add too much confusion.